boa_engine/src/syntax/parser/expression/primary/mod.rs

```diff
@@ -126,17 +126,42 @@ where
                 ClassExpression::new(self.name, self.allow_yield, self.allow_await)
                     .parse(cursor, interner)
             }
-            TokenKind::Keyword((Keyword::Async, false)) => {
-                cursor.next(interner).expect("token disappeared");
-                let mul_peek = cursor.peek(1, interner)?.ok_or(ParseError::AbruptEnd)?;
-                if mul_peek.kind() == &TokenKind::Punctuator(Punctuator::Mul) {
-                    AsyncGeneratorExpression::new(self.name)
-                        .parse(cursor, interner)
-                        .map(Node::from)
+            TokenKind::Keyword((Keyword::Async, contain_escaped_char)) => {
+                let contain_escaped_char = *contain_escaped_char;
+                let is_generator = if let Ok(kind) = cursor.peek_kind(2, interner) {
+                    kind == &TokenKind::Punctuator(Punctuator::Mul)
                 } else {
-                    AsyncFunctionExpression::new(self.name, self.allow_yield)
+                    false
+                };
+
+                // FIXME: even if the next character is OpenParen, it could be an identifier.
+                // e.g.) async();
```

Comment:
Tricky parsing here. Maybe we should assume that it's an identifier, unless we encounter another identifier, for example, afterwards?

Reply:
async lambda function does not have an identifier as a next token. e.g.) `async () => { }`.

Reply:
Makes sense, yes, thanks!

Reply:
filed a bug here: https://github.com/boa-dev/boa/issues/2275